packages/wrangler/src/sites.tsx

```diff
@@ -70,49 +72,93 @@ async function createKVNamespaceIfNotAlreadyExisting(
+ * asset in the KV namespace.
+ */
 export async function syncAssets(
   accountId: string,
   scriptName: string,
-  dirPath: string,
+  siteAssets: AssetPaths | undefined,
   preview: boolean,
   _env?: string
-) {
+): Promise<{ manifest: { [filePath: string]: string }; namespace: string }> {
+  if (siteAssets === undefined) {
+    return { manifest: undefined, namespace: undefined };
+  }
+
   const title = `__${scriptName}_sites_assets${preview ? "_preview" : ""}`;
   const { id: namespace } = await createKVNamespaceIfNotAlreadyExisting(
     title,
     accountId
   );
 
   // let's get all the keys in this namespace
-  const keys = new Set(
-    (await listNamespaceKeys(accountId, namespace)).map((x) => x.name)
-  );
+  const result = await listNamespaceKeys(accountId, namespace);
+  const keys = new Set(result.map((x) => x.name));
 
   const manifest = {};
   const upload: {
     key: string;
     value: string;
     base64: boolean;
   }[] = [];
+
+  const include = createPatternMatcher(siteAssets.includePatterns, false);
+  const exclude = createPatternMatcher(siteAssets.excludePatterns, true);
+
   // TODO: this can be more efficient by parallelising
-  for await (const file of getFilesInFolder(dirPath)) {
-    // TODO: "exclude:" config
-    const { filePath } = await hashFile(file);
+  for await (const file of getFilesInFolder(siteAssets.baseDirectory)) {
+    const relativePath = path.relative(siteAssets.baseDirectory, file);
+    if (!include(relativePath)) {
+      continue;
+    }
+    if (exclude(relativePath)) {
+      continue;
+    }
+
+    await validateAssetSize(file);
+
+    const { assetKey } = await hashAsset(file);
     // now put each of the files into kv
-    if (!keys.has(filePath)) {
+    if (!keys.has(assetKey)) {
       console.log(`uploading ${file}...`);
       const content = await readFile(file, "base64");
-      if (content.length > 25 * 1024 * 1024) {
-        throw new Error(`File ${file} is too big, it should be under 25 mb.`);
-      }
       upload.push({
-        key: filePath,
+        key: assetKey,
         value: content,
         base64: true,
       });
     }
-    manifest[path.relative(dirPath, file)] = filePath;
+    manifest[path.relative(siteAssets.baseDirectory, file)] = assetKey;
   }
   await putBulkKeyValue(accountId, namespace, JSON.stringify(upload));
   return { manifest, namespace };
 }
+
+function createPatternMatcher(
+  patterns: string[],
+  exclude: boolean
+): (filePath: string) => boolean {
+  if (patterns.length === 0) {
+    return (_filePath) => !exclude;
+  } else {
+    const ignorer = ignore().add(patterns);
+    return (filePath) => ignorer.test(filePath).ignored;
+  }
+}
+
+async function validateAssetSize(filePath: string) {
+  const { size } = await stat(filePath);
+  if (size > 25 * 1024 * 1024) {
+    throw new Error(`File ${filePath} is too big, it should be under 25 mb.`);
```

Comment:
This is actually 25 MiB. Calculation is correct, but error message is not.

Reply:
LOL!

Yeah technically it is MiB, but the majority of the world (AFAICT) would be thrown by the use of that over MB, which is commonly the most used term. I think using the word Mebibyte would definitely raise some eyebrows.

https://www.majordifferences.com/2018/03/differences-between-megabyte-and.html

Reply:
For what it is worth Wrangler1 uses `1048576 bytes` in its error message, which is also technically accurate but not very user friendly IMO.

Reply:
Pages has had questions about it. Currently, we're incorrectly capping file sizes at 25 MB and people have pointed to the KV documentation asking why there's a difference (https://developers.cloudflare.com/workers/platform/limits#kv-limits). Our next release will update our limit and error messaging to 25 MiB.

I think most people who are unfamiliar will just read it as 25 MB. Certainly, that's what I used to do 😅

Reply:
Where is the size check in the `pages` code?

Reply:
I ask because we were previously checking the base64 size, not the actual byte size of the asset being uploaded, which would also cause files that are close but under 25MiB to fail to be uploaded.

Reply:
I meant for in the Cloudflare Pages build image (internal code).